rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,248 @@
+- Connecting pipelines together - The ability to take input from another pipeline.
+- Component reuse - The ability to define boilerplate for reuse across many different pipelines. This will likely align with Datadog’s “pipeline catalogue”.
+- Pipeline quotas - The ability to limit how much data a pipeline can send to a sink.
+
+## Pain
+
+- The only possibility to reuse the same set of transform between multiple configuration file is by duplicating them.
+- If an admin wants to split responsibilities between `ops` and `devs` by only allowing the `ops` to deal with `sources/sinks` and the `devs`, it's currently not possible.
+
+## User Experience
+
+These changes are providing a way for vector users to split their configuration in a way that improves the collaboration between ops and devs.
+This split will take be made by creating a pipeline configuration file, in which the devs will be able to configure their components.
+The ops will now configure their `sources` and `sinks`, and expose them to be used by the devs.
+The ops will be able to only given access to some of the common components when configuring the pipeline.
+The devs will have the ability to consume the provided components without being able to change the common configuration.
+The users will be able to monitor their pipelines through the `internal_metrics` source.
+
+## Motivation
+
+- Helps Vector to grow organically within an organization by allowing teams to adopt Vector at their own pace without heavy administrator involvement.
+- Reduces the management overhead of devops/SREs by enabling teams to manage their own pipelines (spread the management load).
+
+## Implementation
+
+### How and where the pipelines are stored
+
+To avoid backward incompatibility, pipelines will be loaded from a `pipelines` sub-directory relative to the Vector configuration directory (e.g., `/etc/vector/pipelines`). Therefore, if a user changes the location of the Vector configuration directory they will also change the `pipelines` directory path. They are coupled.
+The `pipelines` directory will contain all pipelines represented as individual files. For simplicity, and to ensure users do not over complicate pipeline management, sub-directories/nesting are not allowed. This is inspired by Terraform's single-level directory nesting, which has been a net positive for simple management of large Terraform projects.
+Each pipeline file is a processing subset of a larger Vector configuration file. Therefore, it follows the same syntax as Vector's configuration (`toml`, `yaml`, and `json`).
+
+### Configuration
+
+Each time Vector will read its configuration file (on boot and when it live reloads), it will read the referring pipeline configuration files and associate an `id` attribute corresponding to the pipeline's filename.
+
+A pipeline can be referenced in a configuration file as follows
+
+```toml
+# vector.toml
+[resources.pipeline_id]
```

Comment:
In general, I disagree with requiring pipelines to be defined in the "root" `vector.toml` file due to workflow reasons. The root `vector.toml` file will be maintained by an admin (not teams), and there will be circumstances where teams should be able to onboard themselves without admin intervention. Therefore, adding a file in the `pipelines` direction should be enough to onboard a pipeline.

This is not to say we shouldn't allow configuration options for pipelines in the root `vector.toml` file, but they should be optional, and these options should be for admins (restricting what pipelines).

cc @jszwedko since this was his idea.

Reply:
so maybe, to keep @lukesteensen 's logic of keeping things simple, we should keep this restriction part for another RFC?

Reply:
Ok, you answered this after ;)